lib/tracing-limit/src/lib.rs

```diff
@@ -515,187 +756,305 @@ mod test {
+            ]
+        );
+    }
+
+    #[test]
+    #[serial]
+    fn rate_limit_by_span_key() {
+        let (events, sub) = setup_test(1);
         tracing::subscriber::with_default(sub, || {
             for _ in 0..21 {
-                info!(
-                    message = "Hello world!",
-                    internal_log_rate_limit = true,
-                    internal_log_rate_secs = 1
-                );
+                for key in &["foo", "bar"] {
+                    let span = info_span!("span", component_id = &key);
+                    let _enter = span.enter();
+                    info!(message = format!("Hello {key}!").as_str());
+                }
                 MockClock::advance(Duration::from_millis(100));
             }
         });
 
         let events = events.lock().unwrap();
 
+        // Events with different component_id values create separate rate limit groups
         assert_eq!(
             *events,
             vec![
-                "Hello world!",
-                "Internal log [Hello world!] is being suppressed to avoid flooding.",
-                "Internal log [Hello world!] has been suppressed 9 times.",
-                "Hello world!",
-                "Internal log [Hello world!] is being suppressed to avoid flooding.",
-                "Internal log [Hello world!] has been suppressed 9 times.",
-                "Hello world!",
+                event!("Hello foo!", component_id: "foo"),
+                event!("Hello bar!", component_id: "bar"),
+                event!("Internal log [Hello foo!] is being suppressed to avoid flooding.", component_id: "foo"),
+                event!("Internal log [Hello bar!] is being suppressed to avoid flooding.", component_id: "bar"),
+                event!("Internal log [Hello foo!] has been suppressed 9 times.", component_id: "foo"),
+                event!("Hello foo!", component_id: "foo"),
+                event!("Internal log [Hello bar!] has been suppressed 9 times.", component_id: "bar"),
+                event!("Hello bar!", component_id: "bar"),
+                event!("Internal log [Hello foo!] is being suppressed to avoid flooding.", component_id: "foo"),
+                event!("Internal log [Hello bar!] is being suppressed to avoid flooding.", component_id: "bar"),
+                event!("Internal log [Hello foo!] has been suppressed 9 times.", component_id: "foo"),
+                event!("Hello foo!", component_id: "foo"),
+                event!("Internal log [Hello bar!] has been suppressed 9 times.", component_id: "bar"),
+                event!("Hello bar!", component_id: "bar"),
             ]
-            .into_iter()
-            .map(std::borrow::ToOwned::to_owned)
-            .collect::<Vec<String>>()
         );
     }
 
     #[test]
-    fn rate_limit_by_span_key() {
-        let _guard = TRACING_DEFAULT_LOCK.lock().unwrap();
+    #[serial]
+    fn rate_limit_by_event_key() {
```

Comment:
I'd like to see an additional test very similar to this but with the same log message. Something like this:

```
event!("Hello!", component_id: "foo"),
event!("Internal log [Hello!] is being suppressed to avoid flooding.", component_id: "foo"),
// Now we are sure that `foo` is rate limited
event!("Hello!", component_id: "bar"), // but `bar` isn't since it is a different `component_id`
```